### DISCUSSION TITLE
Bluetooth connectivity problems

### DISCUSSION THREAD

**Discussion by simelis** [NightscoutFoundation/xDrip#3165] [Bug reports]
Greetings to everyone, 

My son has been using the latest xdrip+ with dex1 sensor and his Nokia G21 phone for a while, but the phone is experiencing constant Bluetooth connectivity problems. At least once a day Bluetooth gets stuck and xdrip loses BG data. At the same time AAPS issues "Pump is not reachable" notification, although the phone is with him, which indicates it's a global problem with the phone's Bluetooth connectivity. I noticed Bluetooth watchdog doesn't get activated, so I usually resolve the problem by restarting a phone or killing xdrip, manually turning Bluetooth off and on, and then starting xdrip again. That is quite exhausting, as it happens too often and manual actions are required.

I enabled debug logging, looked into the code and noticed Bluetooth watchdog only gets activated when Minimize Scanning option is enabled, and it activates only after about 10 to 12 unsuccessful connection attempts. Isn't that too much? I think that is why I resolve the issue with manual restart before the watchdog gets activated, wouldn't 3-4 retries be enough?

While running through the Ob1G5CollectionService, I also noticed that use_auto_connect  variable is used in a manner which is not clear to me - it never gets changed, and the logic in handleWakeup where it is used is messed up as far as I understand. Shouldn't it go to always_scan mode only after both CONNECT_NOW and CONNECT fail? And after always_scan mode gets activated because of connectivity problems, shouldn't it be deactivated after successful connection or manual collector restart? Now once activated, it remains on until xdrip restarts as far as I understand, is it intended?

I also noticed that those connection problems are usually followed by BleDisconnectedException with status 133 (GATT_ERROR).
I have tried initiating Bluetooth reset when BleDisconnectedException  with status 133 is caught, and the problems seem to be resolved - I get no noticeable data loss so far. I suggest a Bluetooth setting like "Aggresive Bluetooth restart" could enable this behavior.

Does it make sense? Should I create a pull request with the changes I've made? 

Best regards, your effort in making life with DM comfortable is most appreciated!

**Comment by Navid200** [comment]
Why do you have minimize scanning disabled for Nokia G21?  According to my search, it is Android 11 and is upgradable to Android 13.
Does this mean you are not following the recommended settings?
If yes, I'm afraid, opening a PR will not be the best course of action.

Please start by using the recommended settings: https://navid200.github.io/xDrip/docs/G6-Recommended-Settings.html

If the problem persists, and if your modifications to xDrip still improve the outcome with the recommended settings, please open a PR.

If your thinking is that you want to change xDrip because it works for you and why should you care about the recommended settings, please note that xDrip must work for everyone.  If your changes fix a Nokia problem but break xDrip for all the other phones, needless to say, we cannot proceed.

If you open a PR, the worst thing that could happen is that it will not be merged.  But, I don't want you to waste your time if you are not even using the recommended settings.

**Comment by simelis** [comment]
I surely started with recommended settings by following everything in your link, but it was even worse then, so I tried everything, and unchecking it reduced the disconnections to about once a day. Sorry for not making it clear. Anyway, my observations are valid with minimize scanning enabled - I mentioned it only because I noticed it in code, and though it might not be intended.

Yes, the phone is on Android 13.

**Reply by Navid200** [reply]
Are you sure you have disabled battery optimization?
https://dontkillmyapp.com/nokia

**Reply by simelis** [reply]
I am sure I have disabled battery optimization following your recommendations - app battery usage is set to unrestricted for both xdrip and bluetooth system app. Regarding a special Nokia case described in https://dontkillmyapp.com/nokia, I cannot find neither of mentioned apps, so I guess my problem cannot be resolved this way.

**Comment by simelis** [comment]
I created a PR here: https://github.com/NightscoutFoundation/xDrip/pull/3167
